File: Day2/2-1.hs
data Direction = U | R | D | L deriving (Show, Read)
type Command = [Direction]
type Instructions = [Command]

keypadWidth :: Int
keypadWidth = 3
keypadHeight :: Int
keypadHeight = 3
keypadSize :: Int
keypadSize = keypadHeight * keypadWidth

getNumber :: Int -> Command -> Int
getNumber n [] = n
getNumber n (U:cs) = getNumber nextNum cs
                     where potentialMove = n - keypadWidth
                           nextNum = if potentialMove > 0 then potentialMove else n
getNumber n (R:cs) = getNumber nextNum cs
                     where potentialMove = n + 1
                           nextNum = if potentialMove <= keypadSize &&
                                        n `mod` keypadWidth /= 0
                                     then potentialMove
                                     else n
getNumber n (D:cs) = getNumber nextNum cs
                     where potentialMove = n + keypadWidth
                           nextNum = if potentialMove <= keypadSize then potentialMove else n
getNumber n (L:cs) = getNumber nextNum cs
                     where potentialMove = n - 1
                           nextNum = if potentialMove > 0 &&
                                        n `mod` keypadWidth /= 1
                                     then potentialMove
                                     else n

readInstructions :: Int -> Instructions -> [Int]
readInstructions _ [] = []
readInstructions n (i:is) = let num = getNumber n i
                            in num : readInstructions num is

main :: IO ()
main = do
  strInstr <- fmap lines $ readFile "input.txt"
  let instr = map (\a -> map (\b -> read [b] :: Direction) a) strInstr
  putStrLn $ show $ readInstructions 5 instr

File: Day02_JavaScript/2-1.js
var fs = require('fs');

var keypadWidth = 3,
    keypadHeight = 3,
    keypadSize = keypadWidth * keypadHeight;

function moveUp(n) {
    var potentialMove = n - keypadWidth;
    return potentialMove > 0 ? potentialMove : n;
}

function moveDown(n) {
    var potentialMove = n + keypadWidth;
    return potentialMove <= keypadSize ? potentialMove : n;
}

function moveRight(n) {
    var potentialMove = n + 1;
    return potentialMove <= keypadSize && n % 3 !== 0 ? potentialMove : n;
}

function moveLeft(n) {
    var potentialMove = n - 1;
    return potentialMove > 0 && n % 3 !== 1 ? potentialMove : n;
}

function getNumber(n, commands) {
    var num = n;
    for (var i = 0, len = commands.length; i < len; i++) {
        var command = commands.charAt(i);
        if (command == 'U') num = moveUp(num);
        else if (command == 'D') num = moveDown(num);
        else if (command == 'R') num = moveRight(num);
        else if (command == 'L') num = moveLeft(num);
    }
    return num;
}

function readInstructions(instructions) {
    var code = "";
    var num = 5;
    for (var i = 0, len = instructions.length; i < len; i++) {
        num = getNumber(num, instructions[i]);
        code += parseInt(num);
    }
    return code;
}

function solveCodeFromInput(input) {
    var lines = input.match(/[^\r\n]+/g);
    return readInstructions(lines);
}

fs.readFile('input.txt', 'utf8', function(err, input) {
    if (err)
        console.log('Could not read input file.');
    else
        console.log(solveCodeFromInput(input));
});
